src/core/metadata/qgslayermetadata.cpp

```diff
@@ -273,6 +278,312 @@ void QgsLayerMetadata::readFromLayer( const QgsMapLayer *layer )
+      oneAddress.type = addressElement.namedItem( QStringLiteral( "type" ) ).toElement().text();
+      addresses << oneAddress ;
+    }
+    oneContact.addresses = addresses;
+    addContact( oneContact );
+  }
+
+  // links
+  mnl = metadataElement.namedItem( QStringLiteral( "links" ) );
+  mne = mnl.toElement();
+  QDomNodeList el = mne.elementsByTagName( QStringLiteral( "link" ) );
+  for ( int i = 0; i < el.size(); i++ )
+  {
+    mne = el.at( i ).toElement();
+    QgsLayerMetadata::Link oneLink;
+    oneLink.name = mne.attribute( QStringLiteral( "name" ) );
+    oneLink.type = mne.attribute( QStringLiteral( "type" ) );
+    oneLink.url = mne.attribute( QStringLiteral( "url" ) );
+    oneLink.description = mne.attribute( QStringLiteral( "description" ) );
+    oneLink.format = mne.attribute( QStringLiteral( "format" ) );
+    oneLink.mimeType = mne.attribute( QStringLiteral( "mimeType" ) );
+    oneLink.size = mne.attribute( QStringLiteral( "size" ) );
+    addLink( oneLink );
+  }
+
+  return true;
+}
+
+bool QgsLayerMetadata::writeMetadataXml( QDomElement &metadataElement, QDomDocument &document ) const
+{
+  // identifier
+  QDomElement identifier = document.createElement( QStringLiteral( "identifier" ) );
+  QDomText identifierText = document.createTextNode( mIdentifier );
+  identifier.appendChild( identifierText );
+  metadataElement.appendChild( identifier );
+
+  // parent identifier
+  QDomElement parentIdentifier = document.createElement( QStringLiteral( "parentidentifier" ) );
+  QDomText parentIdentifierText = document.createTextNode( mParentIdentifier );
+  parentIdentifier.appendChild( parentIdentifierText );
+  metadataElement.appendChild( parentIdentifier );
+
+  // language
+  QDomElement language = document.createElement( QStringLiteral( "language" ) );
+  QDomText languageText = document.createTextNode( mLanguage );
+  language.appendChild( languageText );
+  metadataElement.appendChild( language );
+
+  // type
+  QDomElement type = document.createElement( QStringLiteral( "type" ) );
+  QDomText typeText = document.createTextNode( mType );
+  type.appendChild( typeText );
+  metadataElement.appendChild( type );
+
+  // title
+  QDomElement title = document.createElement( QStringLiteral( "title" ) );
+  QDomText titleText = document.createTextNode( mTitle );
+  title.appendChild( titleText );
+  metadataElement.appendChild( title );
+
+  // abstract
+  QDomElement abstract = document.createElement( QStringLiteral( "abstract" ) );
+  QDomText abstractText = document.createTextNode( mAbstract );
+  abstract.appendChild( abstractText );
+  metadataElement.appendChild( abstract );
+
+  // keywords
+  QMapIterator<QString, QStringList> i( keywords() );
+  while ( i.hasNext() )
+  {
+    i.next();
+    QDomElement keywordsElement = document.createElement( QStringLiteral( "keywords" ) );
+    keywordsElement.setAttribute( QStringLiteral( "vocabulary" ), i.key() );
+    for ( const QString kw : i.value() )
```

Comment:
This will definitely detach. It needs to be:

    const QStringList values = i.value();
    for ( const QString& kw : values )

Note also the missing reference on kw